---
title: "documentation"
editor: visual
---

The nf-core/pathogensurveillance pipeline comes with documentation about the pipeline [usage](https://nf-co.re/pathogensurveillance/usage), [parameters](https://nf-co.re/pathogensurveillance/parameters) and [output](https://nf-co.re/pathogensurveillance/output).

### Input format

The primary input to the pipeline is a CSV (comma comma-separated value). Columns can be in any order and unneeded columns can be left out or left blank. Only a single column containing paths to raw sequence data or SRA (Sequence Read Archive) accessions is required and each sample can have values in different columns. Any columns not recognized by `pathogensurveillance` will be ignored, allowing users to adapt existing sample metadata table by adding new columns. Below is a description of each column used by `pathogensurveillance`:

-   **sample_id**: The unique identifier for each sample. This will be used in file names to distinguish samples in the output. Each sample ID must correspond to a single set of sequence data (The `shortread_*`, `nanopore`, and `sra` columns), although the same sequence data can be used by multiple different IDs. Any values that correspond to multiple different sets of sequence data or contain characters that cannot appear in file names (/:\*?"\<\>\| .) will be modified automatically. If not supplied, it will be inferred from the names of input data.
-   **sample_name**: A human-readable label for the sample that is used in plots and tables. If not supplied, it will be inferred from the names of `sample_id`.
-   **shortread_1**: Path to short read FASTQs like that produced by Illumina. When paired end sequencing is used, this is used for the forward read's data. This can be a local file path or a URL to an online location.
-   **shortread_2**: Path to short read FASTQs like that produced by Illumina. This is used for the reverse read's data when paired-end sequencing is used. This can be a local file path or a URL to an online location.
-   **nanopore**: Path to nanopore FASTQs. This can be a local file path or a URL to an online location.
-   **pacbio**: Path to pacbio FASTQs. This can be a local file path or a URL to an online location.
-   **sra**: Sequence Read Archive (SRA) accession numbers. These will be automatically downloaded and used as input.
-   **reference_id**: The unique identifier for each user-defined reference genome. This will be used in file names to distinguish samples in the output. Each reference ID must correspond to a single set of reference data (The `reference` and `reference_refseq` columns), although the same reference data can be used by multiple different IDs. Any values that correspond to multiple different sets of reference data or contain characters that cannot appear in file names (/:\*?"\<\>\| .) will be modified automatically. If not supplied, it will be inferred from the names of reference genomes.
-   **reference_name**: A human-readable label for user-defined reference genomes that is used in plots and tables. If not supplied, it will be inferred from the names of reference_id.
-   **reference**: Path to user-defined reference genomes for each sample. This can be a local file path or a URL to an online location.
-   **reference_refseq**: RefSeq accession ID for a user-defined reference genome. These will be automatically downloaded and used as input.
-   **report_group**: How to group samples into reports. For every unique value in this column a report will be generated. Samples can be assigned to multiple reports by separating group IDs by `;`. For example `all;subset` will put the sample in both `all` and `subset` report groups. If not included, all samples will be
-   **color_by**: The names of user-specific columns (not usually any of the ones described here) containing variables to base the color of some plots by (e.g. the minimum spanning network). If not included, plots will not be colored.

### Command-line options

**Required:**

-   **--input:** Path to comma-separated file containing information about the samples in the experiment.

-   **--output:** The output directory where the results will be saved. You have to use absolute paths to storage on Cloud infrastructure.

-   **--bakta_db:** The path to the Bakta database folder. This or `--download_bakta_db` must be included.

-   **--download_bakta_db:** Download the database required for running Bakta. Note that this will download gigabytes of information, so if you plan on running the pipeline repeatedly it would be better to download the database manually and specify the path with `--bakta_db`

**Nextflow Options:**

-   **-profile:** Instructs the pipeline to use the named tool for software management. `docker`, `singularity`, `podman`, `shifter`, `charliecloud` and `conda`. For example, `-profile test,docker`
-   **-resume:** Restarts an incomplete run by using cached intermediate files.

**Optional:**

-   **--email:** Set this parameter to your e-mail address to get a summary e-mail with details of the run sent to you when the workflow exits. If set in your user config file (\`\~/.nextflow/config\`) then you don't need to specify this on the command line for every run.

-   **--multiqc_title:** MultiQC report title. Printed as page header, used for filename if not otherwise specified.

**Performance Parameters:**

-   **--max_cpus:** Maximum number of CPUs that can be requested for any single job. Default: `16`

-   **--max_memory:** Maximum amount of memory that can be requested for any single job. Default: `128.GB`

-   **--max_time:** Maximum amount of time that can be requested for any single job. Default: `240.h`

**Analysis Parameters:**

-   **--sketch_max_depth:** Depth reads are subsampled to for the initial sketch-based identification. Default: `3`

-   **--variant_max_depth:** Depth reads are subsampled to for the variant-based parts of the analysis. Default: `15`

-   **--assembly_max_depth:** Depth reads are subsampled to for genome assembly. This will be multiplied by the predicted ploidy of each sample. Default: `30`

-   **--refseq_download_num:** The maximum number of RefSeq sequences to select and download for each sample at each taxonomic level (species, genus, and family). The total number will vary based on the diversity of samples. Default: `10`

-   **--min_core_genes:** The minimum number of genes needed to conduct a core gene phylogeny. Samples and references will be removed (as allowed by the `min_core_samps` and `min_core_refs` options) until this minimum is met. Default: `10`

-   **--min_core_samps:** The minimum proportion of samples needed to conduct a core gene phylogeny. Samples will be removed until the `min_core_genes` option is satisfied or this minimum is met. Default: `0.8`

-   **--min_core_refs:** The minimum proportion of references needed to conduct a core gene phylogeny. References will be removed until the `min_core_genes` option is satisfied or this minimum is met. Default: `0.5`

-   **--max_core_genes:** The maximum number of genes used to conduct a core gene phylogeny. Default: `100`

-   **--min_ref_ani:** The minimum ANI between a sample and potential reference for that reference to be used for variant calling with that sample. To force all the samples in a report group to use the same reference, set this value very low. Default: `0.9`

-   **--copymode:** Storage management setting to determine which files will be copied from the cache into the output directory.

    -   `high` - All files are copied into output directory.

    -   `medium` - Reports are copied. Large sequencing files are not copied, but they are accessible through symlinks to their location in the cache (default).

    -   `low` - No files are copied into output directory, but files are accessible through symlinks to their location in the cache.

## Benchmarks {#benchmarks}
