---
title: "faq"
editor: visual
---

*How long does the pathogensurveillance take to run?*

*How many samples can be run at once?*

*What are the minimum computational resources necessary to run pathogensurveillance?*

*How much sequencing coverage is needed?*

*Can pathogensurveillance be run on macOS?*

How are reference genomes selected?

This process works remarkably well for most use cases. However, there are some biological reasons why selecting a reference genome becomes a more challenging problem (with or without pathogensurveilance). First, if you are working with a truly unique organism, there may be no nearby reference. Second, closely related bacteria can still vary considerably their accessory genomes. This process will prioritze matches between core genes above other genomic content, like extra-chromosomal plasmids or other rapidly evolving/repetitive regions. If these regions are of diagnostic interest for your system, you may get better results if you are able to specify a known reference before running the pipeline.
