<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Tutorial</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<link rel="shortcut icon" href="images/favicon.png">


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./tutorial.html">Tutorial</a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Pathogen Surveillance Documentation</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Home</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./quickstart.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Quickstart</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./documentation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Documentation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Tutorial</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./examplereports.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Example Reports</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./citations.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Citations</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#example-1-standard-run" id="toc-example-1-standard-run" class="nav-link active" data-scroll-target="#example-1-standard-run">Example 1: Standard Run</a>
  <ul class="collapse">
  <li><a href="#sample-input" id="toc-sample-input" class="nav-link" data-scroll-target="#sample-input">Sample input</a></li>
  <li><a href="#running-the-pipeline" id="toc-running-the-pipeline" class="nav-link" data-scroll-target="#running-the-pipeline">Running the pipeline</a></li>
  <li><a href="#report" id="toc-report" class="nav-link" data-scroll-target="#report">Report</a></li>
  </ul></li>
  <li><a href="#example-2-defining-references" id="toc-example-2-defining-references" class="nav-link" data-scroll-target="#example-2-defining-references">Example 2: Defining References</a>
  <ul class="collapse">
  <li><a href="#specifying-primary-references" id="toc-specifying-primary-references" class="nav-link" data-scroll-target="#specifying-primary-references">specifying primary references</a></li>
  <li><a href="#specifying-contextual-references" id="toc-specifying-contextual-references" class="nav-link" data-scroll-target="#specifying-contextual-references">specifying contextual references</a></li>
  <li><a href="#selecting-references-from-an-ncbi-query" id="toc-selecting-references-from-an-ncbi-query" class="nav-link" data-scroll-target="#selecting-references-from-an-ncbi-query">selecting references from an ncbi query</a></li>
  <li><a href="#multiple-references-per-sample" id="toc-multiple-references-per-sample" class="nav-link" data-scroll-target="#multiple-references-per-sample">Multiple references per sample</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Tutorial</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="example-1-standard-run" class="level2">
<h2 class="anchored" data-anchor-id="example-1-standard-run">Example 1: Standard Run</h2>
<p>This example uses sequencing reads from an outbreak of <em>Xanthomonas hortorum</em> in several plant nurseries. We’ll be treating the pathogen as an unknown and using the pathogensurveillance pipeline to determine what we know already (that these samples come from <em>Xanthomonas hortorum</em>). We’ll also explore how isolates from different nursery populations relate to each other and the reference sequences of other closely-related organisms. This information can be obtained from several plots that the pathogensurveillance pipeline generates automatically. <br><br></p>
<section id="sample-input" class="level3">
<h3 class="anchored" data-anchor-id="sample-input">Sample input</h3>
<p>The pipeline is designed to work with a wide variety of existing metadata sheets without extensive changes. Here’s a look at “xanthomonas.csv”, which serves as the only unique input file within the command to run the pipeline:</p>
<div class="cell">
<div class="cell-output-display">
<div class="kable-table">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 4%">
<col style="width: 17%">
<col style="width: 17%">
<col style="width: 4%">
<col style="width: 6%">
<col style="width: 8%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 2%">
<col style="width: 11%">
<col style="width: 3%">
<col style="width: 3%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">sample_id</th>
<th style="text-align: left;">shortread_1</th>
<th style="text-align: left;">shortread_2</th>
<th style="text-align: left;">reference</th>
<th style="text-align: left;">reference_id</th>
<th style="text-align: left;">report_group</th>
<th style="text-align: left;">color_by</th>
<th style="text-align: left;">date_isolated</th>
<th style="text-align: left;">date_received</th>
<th style="text-align: right;">year</th>
<th style="text-align: left;">host</th>
<th style="text-align: left;">cv_key</th>
<th style="text-align: left;">nursery</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">22-299</td>
<td style="text-align: left;">test/data/reads/22-299_R1.fastq.gz</td>
<td style="text-align: left;">test/data/reads/22-299_R2.fastq.gz</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">xan_test;subgroup</td>
<td style="text-align: left;">year;nursery</td>
<td style="text-align: left;">3/2/22</td>
<td style="text-align: left;">3/29/22</td>
<td style="text-align: right;">2022</td>
<td style="text-align: left;">Pelargonium x hortorum</td>
<td style="text-align: left;">CV-1</td>
<td style="text-align: left;">MD</td>
</tr>
<tr class="even">
<td style="text-align: left;">22-300</td>
<td style="text-align: left;">test/data/reads/22-300_R1.fastq.gz</td>
<td style="text-align: left;">test/data/reads/22-300_R2.fastq.gz</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">xan_test;subgroup</td>
<td style="text-align: left;">year;nursery</td>
<td style="text-align: left;">3/2/22</td>
<td style="text-align: left;">3/30/22</td>
<td style="text-align: right;">2022</td>
<td style="text-align: left;">Pelargonium x hortorum</td>
<td style="text-align: left;">CV-2</td>
<td style="text-align: left;">MD</td>
</tr>
<tr class="odd">
<td style="text-align: left;">22-301</td>
<td style="text-align: left;">test/data/reads/22-301_R1.fastq.gz</td>
<td style="text-align: left;">test/data/reads/22-301_R2.fastq.gz</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">xan_test;subgroup</td>
<td style="text-align: left;">year;nursery</td>
<td style="text-align: left;">3/2/22</td>
<td style="text-align: left;">3/31/22</td>
<td style="text-align: right;">2022</td>
<td style="text-align: left;">Pelargonium x hortorum</td>
<td style="text-align: left;">CV-3</td>
<td style="text-align: left;">MD</td>
</tr>
<tr class="even">
<td style="text-align: left;">22-302</td>
<td style="text-align: left;">test/data/reads/22-302_R1.fastq.gz</td>
<td style="text-align: left;">test/data/reads/22-302_R2.fastq.gz</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">xan_test;subgroup</td>
<td style="text-align: left;">year;nursery</td>
<td style="text-align: left;">3/2/22</td>
<td style="text-align: left;">4/1/22</td>
<td style="text-align: right;">2022</td>
<td style="text-align: left;">Pelargonium x hortorum</td>
<td style="text-align: left;">CV-4</td>
<td style="text-align: left;">MD</td>
</tr>
<tr class="odd">
<td style="text-align: left;">22-303</td>
<td style="text-align: left;">test/data/reads/22-303_R1.fastq.gz</td>
<td style="text-align: left;">test/data/reads/22-303_R2.fastq.gz</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">xan_test;subgroup</td>
<td style="text-align: left;">year;nursery</td>
<td style="text-align: left;">3/2/22</td>
<td style="text-align: left;">4/2/22</td>
<td style="text-align: right;">2022</td>
<td style="text-align: left;">Pelargonium x hortorum</td>
<td style="text-align: left;">CV-5</td>
<td style="text-align: left;">MD</td>
</tr>
<tr class="even">
<td style="text-align: left;">22-304</td>
<td style="text-align: left;">test/data/reads/22-304_R1.fastq.gz</td>
<td style="text-align: left;">test/data/reads/22-304_R2.fastq.gz</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
<td style="text-align: left;">xan_test;subgroup</td>
<td style="text-align: left;">year;nursery</td>
<td style="text-align: left;">3/7/22</td>
<td style="text-align: left;">4/3/22</td>
<td style="text-align: right;">2022</td>
<td style="text-align: left;">Pelargonium x hortorum</td>
<td style="text-align: left;">CV-6</td>
<td style="text-align: left;">MD</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<p>There is quite a bit of information in this file, but only a few columns are essential (and can be in any order). The input csv needs show the pipeline where to find the sequencing reads. These can be present either locally or they can be downloaded.</p>
<p><strong>Using local reads</strong>: Columns “shortread_1” and “shortread_2” specify the path to forward and reverse reads. Each row corresponds to one individual sample. Reads for this tutorial are hosted on the pathogensurveilance github repo. They are derived from paired-end illumina shortreads, but the pipeline will also work with mixed inputs of Pacbio or Oxford Nanopore sequences.</p>
<p><strong>Downloading reads</strong>: Sequence files may instead be hosted on the ncbi. In that case, the “shortread_1/shortread_2” columns should be substituted with a single “SRA” column, and they will be downloaded from the ncbi automatically. See test/data/metadata/xanthomonas.csv for an example using this input format.</p>
<p><strong>Specifying a reference genome (optional)</strong>: The “reference_refseq” column may be useful when you are relatively confident as to the identity of your samples and would like to include one particular reference for comparison. See Documentation for an in-depth explanation of how to designate mandatory and optional references.</p>
<p><strong>Assigning sample groups (optional)</strong>: The optional column “color_by” is used for data visualization. It will assign one or more columns to serve as grouping factors for the output report. Here, samples will be grouped by the values of the “year” and “nursery” columns. Note that multiple factors need to be separated by semicolons within the color_by column. <br><br></p>
</section>
<section id="running-the-pipeline" class="level3">
<h3 class="anchored" data-anchor-id="running-the-pipeline">Running the pipeline</h3>
<p>Here is the full command used execute this example, using a docker container:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">nextflow</span> run nf-core/pathogensurveillance <span class="at">--input</span> https://raw.githubusercontent.com/grunwaldlab/pathogensurveillance/master/test/data/metadata/xanthomonas.csv <span class="at">--outdir</span> xanthomonas <span class="at">--download_bakta_db</span> true <span class="at">-profile</span> docker <span class="at">-resume</span> <span class="at">--max_cpus</span> 8 <span class="at">--max_memory</span> 30GB <span class="at">-resume</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>When running your own analysis, you will need to provide your own path to the input CSV file.</p>
<p>By default, the pipeline will run on 128 GB of RAM and 16 threads. This is more resources than is strictly necessary and beyond the capacity of most desktop computers. We can scale this back a bit for this lightweight test run. This analysis will work with 8 cpus and 30 GB of RAM (albeit more slowly), which is specified by the –max_cpus and –max_memory settings.</p>
<p>The setting <code>-resume</code> is only necessary when resuming a previous analysis. However, it doesn’t hurt to include it at the start. If the pipeline is interrupted, this setting allows progress to pick up where it left off – as long as the previous command is executed from the same working directory.</p>
<p>If the pipeline begins successfully, you should see a screen tracking your progress:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">[25/63dcee]</span> process <span class="op">&gt;</span> PATHOGENSURVEILLANCE:INPUT_CHECK:SAMPLESHEET_CHECK <span class="er">(</span><span class="ex">xanthomonas.csv</span><span class="kw">)</span><span class="ex">[100%]</span> 1 of 1</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">[-</span>        ] process <span class="op">&gt;</span> PATHOGENSURVEILLANCE:SRATOOLS_FASTERQDUMP                              <span class="at">-</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="ex">[-</span>        ] process <span class="op">&gt;</span> PATHOGENSURVEILLANCE:DOWNLOAD_ASSEMBLIES                               <span class="at">-</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="ex">[-</span>        ] process <span class="op">&gt;</span> PATHOGENSURVEILLANCE:SEQKIT_SLIDING                                    <span class="at">-</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="ex">[-</span>        ] process <span class="op">&gt;</span> PATHOGENSURVEILLANCE:FASTQC                                            <span class="at">-</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="ex">[-</span>        ] process <span class="op">&gt;</span> PATHOGENSURVEILLANCE:COARSE_SAMPLE_TAXONOMY:BBMAP_SENDSKETCH           <span class="at">-</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The input and output of each process can be accessed from the work/ directory. The subdirectory within work/ is designated by the string to left of each step. Note that this location will be different each time the pipeline is run, and only the first part of the name of the subdirectory is shown. For this run, we could navigate to <code>work/25/63dcee(etc)</code> to access the input csv that is used for the next step. <br><br></p>
</section>
<section id="report" class="level3">
<h3 class="anchored" data-anchor-id="report">Report</h3>
<p>You should see a message similar to this if the pipeline finishes successfully:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">-[nf-core/plantpathsurveil]</span> Pipeline completed successfully-</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">To</span> clean the cache, enter the command: </span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="ex">nextflow</span> clean evil_boyd <span class="at">-f</span> </span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="ex">Completed</span> at: 20-May-2024 12:44:40</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="ex">Duration</span>    : 3h 29m 2s</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="ex">CPU</span> hours   : 15.2</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="ex">Succeeded</span>   : 253</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The final report can be viewed as either a .pdf or .html file. It can be accessed inside the reports folder of the output directory (here: <code>xanthomonas/reports</code>). This report shows several key pieces of information about your samples.</p>
<p>A note on storage management - pathogensurveillance creates a large number of intermediate files. For most users we recommend clearing these files after each run. To do so, run the script shown after the completion message (nextflow clean <your_run_name> -f). You would not want to do this if: (1) You still need to use the caching system. For example, imagine you would like to compare a new sample to 10 samples from a previous run. In that case some files could be reused to make the pipeline work more quickly. (2) You would like to use intermediate files for your own analysis. By default, these files are saved in the output directory as symlinks to their location in the work/ directory, so you would need to retrieve these before clearing the cache. You could use alternatively use the option –copymode high to save all intermediate files to the published directory, though in the short term this doubles the storage footprint of each run.</your_run_name></p>
<p>This particular report has been included as an <a href="https://grunwaldlab.github.io/pathogensurveillance_documentation/xanthomonas" target="_blank">example </a></p>
<hr>
<p><strong>Summary:</strong></p>
<ul>
<li><strong>Pipeline Status Report</strong>: error messages for samples or sample groups</li>
<li><strong>Input Data</strong>: Data read from the input .csv file</li>
</ul>
<hr>
<p><strong>Identification:</strong></p>
<ul>
<li><p><strong>Initial identification</strong>: Coarse identification from the bbmap sendsketch step. The first tab shows best species ID for each sample<em>.</em> The second tab shows similarity metrics between sample sequences and other reference genomes: %ANI (average nucleotide identity), %WKID (weighted kmer identity), and %completeness.</p>
<ul>
<li>For more information about each metric, click the <strong>About this table</strong> tab underneath.</li>
</ul></li>
</ul>
<hr>
<p><img src="images/report_ani_heatmap.png" class="img-fluid"></p>
<p><strong>Most similar organisms</strong>: Shows relationships between samples and references using % ani and % pocp (percentage of conserved proteins). For better resolution, you can interactively zoom in/out of plots.</p>
<hr>
<p><img src="images/report_core_gene_phylogeny.png" class="img-fluid"></p>
<p><strong>Core gene phylogeny</strong>: A core gene phylogeny uses the sequences of all gene shared by all of the genomes included in the tree to infer evolutionary relationships. It is the most robust identification provided by this pipeline, but its precision is still limited by the availability of similar reference sequences. Methods to generate this tree differ between prokaryotes and eukaryotes. Our input to the pipeline was prokaryotic DNA sequences, and the method to build this tree is based upon many different core genes shared between samples and references (for eukaryotes, this is constrained to BUSCO genes). This tree is built with iqtree and based upon shared core genes analyzed using the program pirate. You can highlight branches by hovering over and clicking on nodes.</p>
<hr>
<p><img src="images/report_snp_tree.png" class="img-fluid"></p>
<ul>
<li><strong>SNP trees</strong>: This tree is better suited for visualizing the genetic diversity among samples. However, the core gene phylogeny provides a much better source of information for evolutionary differences among samples and other known references.</li>
</ul>
<hr>
<p><strong>Minimum spanning network</strong></p>
<p><img src="images/Xan_MSN_0.01.png" class="img-fluid"></p>
<p><strong>Minimum spanning network</strong>: The nodes represent unique multilocus genotypes, and the size of nodes is proportional to the # number of samples that share the same genotype. The edges represent the SNP differences between two given genotypes, and the darker the color of the edges, the fewer SNP differences between the two.</p>
<hr>
</section>
</section>
<section id="example-2-defining-references" class="level2">
<h2 class="anchored" data-anchor-id="example-2-defining-references">Example 2: Defining References</h2>
<p>You may already know what your samples are. If so, you may also know the best reference genome and want to tell the pipeline to use it. Other users may have a few different organisms of interest that they want to use as a points of comparison. For example, maybe there is a particularly nasty strain of <em>V. cholerae</em> that you want to see in relation to your other samples. There are a few options to select (or not select) reference genomes for these cases.</p>
<p>Pathogensurveilance has two different categories of reference genomes. Primary references are used for alignment and will always be displayed in phylogenetic trees. In contrast, contextual references are selected before the primary reference is known and may not be used later on in the pipeline. Some contextual references are chosen because they are really close matches to your samples, and these may be selected to become primary references. However, pathogensurveilance will select a few distantly related contextual references too. Some of these are used to “fill out” the phylogeny, and you may want a higher or lower number of contextual references depending on how you want your phylogenetic trees to look.</p>
<section id="specifying-primary-references" class="level3">
<h3 class="anchored" data-anchor-id="specifying-primary-references">specifying primary references</h3>
<p>Take this sample list containing three <em>Mycobacterium abscessus</em> samples and three <em>Mycobacterium leprae</em> samples:</p>
<div class="cell">
<div class="cell-output-display">
<div class="kable-table">
<table class="table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;">sample_id</th>
<th style="text-align: left;">ncbi.accession</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">mycobacterium_abscessus1</td>
<td style="text-align: left;">ERR7253671</td>
</tr>
<tr class="even">
<td style="text-align: left;">mycobacterium_abscessus2</td>
<td style="text-align: left;">ERR7253669</td>
</tr>
<tr class="odd">
<td style="text-align: left;">mycobacterium_abscessus3</td>
<td style="text-align: left;">ERR7253671</td>
</tr>
<tr class="even">
<td style="text-align: left;">mycobacterium_leperae1</td>
<td style="text-align: left;">SRR6241707</td>
</tr>
<tr class="odd">
<td style="text-align: left;">mycobacterium_leperae2</td>
<td style="text-align: left;">SRR6241708</td>
</tr>
<tr class="even">
<td style="text-align: left;">mycobacterium_leperae3</td>
<td style="text-align: left;">SRR6241709</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<p>To force the pipeline to use the NCBI specified <em>Mycobacterium abscessus</em> reference genome for the three <em>Mycobacterium abscessus</em> samples, and likewise make the three <em>Mycobacterium leprae</em> samples use the NCBI specified <em>Mycobacterium leprae</em> genome, we need to tell pathogenserveilance where to find the reference sequences and how to use them. We can either specify a local path to the reads, or this can instead be specified through the ref_ncbi_accession column. Here, how the references are used here is controlled by the ref_primary_usage column:</p>
<div class="cell">
<div class="cell-output-display">
<div class="kable-table">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 32%">
<col style="width: 19%">
<col style="width: 24%">
<col style="width: 23%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">sample_id</th>
<th style="text-align: left;">ncbi.accession</th>
<th style="text-align: left;">ref_ncbi_accession</th>
<th style="text-align: left;">ref_primary_usage</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">mycobacterium_abscessus1</td>
<td style="text-align: left;">ERR7253671</td>
<td style="text-align: left;">GCF_001632805.1</td>
<td style="text-align: left;">required</td>
</tr>
<tr class="even">
<td style="text-align: left;">mycobacterium_abscessus2</td>
<td style="text-align: left;">ERR7253669</td>
<td style="text-align: left;">GCF_001632805.1</td>
<td style="text-align: left;">required</td>
</tr>
<tr class="odd">
<td style="text-align: left;">mycobacterium_abscessus3</td>
<td style="text-align: left;">ERR7253671</td>
<td style="text-align: left;">GCF_001632805.1</td>
<td style="text-align: left;">required</td>
</tr>
<tr class="even">
<td style="text-align: left;">mycobacterium_leprae1</td>
<td style="text-align: left;">SRR6241707</td>
<td style="text-align: left;">GCF_003253775.1</td>
<td style="text-align: left;">required</td>
</tr>
<tr class="odd">
<td style="text-align: left;">mycobacterium_leprae2</td>
<td style="text-align: left;">SRR6241708</td>
<td style="text-align: left;">GCF_003253775.1</td>
<td style="text-align: left;">required</td>
</tr>
<tr class="even">
<td style="text-align: left;">mycobacterium_leprae3</td>
<td style="text-align: left;">SRR6241709</td>
<td style="text-align: left;">GCF_003253775.1</td>
<td style="text-align: left;">required</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<hr>
</section>
<section id="specifying-contextual-references" class="level3">
<h3 class="anchored" data-anchor-id="specifying-contextual-references">specifying contextual references</h3>
<p>Taking the previous <em>Mycobacterium abscessus/leprae</em> example, imagine we would like to see the comparison between <em>Mycobacterium leprae</em> and <em>Mycobacterium tuberculosis</em>. We can do this by including <em>Mycobacterium tuberculosis</em> as a mandatory contextual reference:</p>
<div class="cell">
<div class="cell-output-display">
<div class="kable-table">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 31%">
<col style="width: 18%">
<col style="width: 23%">
<col style="width: 26%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">sample_id</th>
<th style="text-align: left;">ncbi.accession</th>
<th style="text-align: left;">ref_ncbi_accession</th>
<th style="text-align: left;">ref_contextual_usage</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">mycobacterium_abscessus1</td>
<td style="text-align: left;">ERR7253671</td>
<td style="text-align: left;">GCF_001632805.1</td>
<td style="text-align: left;">required</td>
</tr>
<tr class="even">
<td style="text-align: left;">mycobacterium_abscessus2</td>
<td style="text-align: left;">ERR7253669</td>
<td style="text-align: left;">GCF_001632805.1</td>
<td style="text-align: left;">required</td>
</tr>
<tr class="odd">
<td style="text-align: left;">mycobacterium_abscessus3</td>
<td style="text-align: left;">ERR7253671</td>
<td style="text-align: left;">GCF_001632805.1</td>
<td style="text-align: left;">required</td>
</tr>
<tr class="even">
<td style="text-align: left;">mycobacterium_leprae1</td>
<td style="text-align: left;">SRR6241707</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">mycobacterium_leprae2</td>
<td style="text-align: left;">SRR6241708</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">mycobacterium_leprae3</td>
<td style="text-align: left;">SRR6241709</td>
<td style="text-align: left;"></td>
<td style="text-align: left;"></td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<hr>
</section>
<section id="selecting-references-from-an-ncbi-query" class="level3">
<h3 class="anchored" data-anchor-id="selecting-references-from-an-ncbi-query">selecting references from an ncbi query</h3>
<p>It is also possible to submit a valid NCBI query to the pipeline, with reference genomes selected from query hits. For example, if you wanted to test how your Mycobacterium leprae samples compared to a bunch of different other Mycobacterium leprae genomes, your reference csv file may look like:</p>
<div class="cell">
<div class="cell-output-display">
<div class="kable-table">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 31%">
<col style="width: 18%">
<col style="width: 26%">
<col style="width: 23%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">sample_id</th>
<th style="text-align: left;">ncbi.accession</th>
<th style="text-align: left;">ref_ncbi_query</th>
<th style="text-align: right;">ref_ncbi_query_max</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">mycobacterium_abscessus1</td>
<td style="text-align: left;">ERR7253671</td>
<td style="text-align: left;"></td>
<td style="text-align: right;">NA</td>
</tr>
<tr class="even">
<td style="text-align: left;">mycobacterium_abscessus2</td>
<td style="text-align: left;">ERR7253669</td>
<td style="text-align: left;"></td>
<td style="text-align: right;">NA</td>
</tr>
<tr class="odd">
<td style="text-align: left;">mycobacterium_abscessus3</td>
<td style="text-align: left;">ERR7253671</td>
<td style="text-align: left;"></td>
<td style="text-align: right;">NA</td>
</tr>
<tr class="even">
<td style="text-align: left;">mycobacterium_leprae1</td>
<td style="text-align: left;">SRR6241707</td>
<td style="text-align: left;">mycobacterium leprae</td>
<td style="text-align: right;">100</td>
</tr>
<tr class="odd">
<td style="text-align: left;">mycobacterium_leprae2</td>
<td style="text-align: left;">SRR6241708</td>
<td style="text-align: left;">mycobacterium leprae</td>
<td style="text-align: right;">100</td>
</tr>
<tr class="even">
<td style="text-align: left;">mycobacterium_leprae3</td>
<td style="text-align: left;">SRR6241709</td>
<td style="text-align: left;">mycobacterium leprae</td>
<td style="text-align: right;">100</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<p>Some things to keep in mind:</p>
<ul>
<li>Depending on your organism, this may a massive amount of data. <strong>Make sure you have queried NCBI beforehand to get a good handle on how many references you are downloading</strong>.</li>
<li>The optional parameter <code>ref_ncbi_query_max</code> is a good way of limiting this number when you are sampling from a densely populated clade, such as <em>Mycobacterium leprae</em>. This parameter can either be a set number (like shown here) or a percentage.</li>
<li>The NCBI API will fail if there are too many requests. See <a href="https://support.nlm.nih.gov/"><code>ncbi support</code></a> for more detail.</li>
</ul>
<hr>
</section>
<section id="multiple-references-per-sample" class="level3">
<h3 class="anchored" data-anchor-id="multiple-references-per-sample">Multiple references per sample</h3>
<p>If we would like to add multiple references per sample, we can enter this information through a separate reference csv. In this example, we specify one primary reference each for <em>Mycobacterium abscessus</em> and <em>Mycobacterium leprae</em>, then three additional contextual references for <em>Mycobacterium leprae</em>:</p>
<div class="cell">
<div class="cell-output-display">
<div class="kable-table">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 13%">
<col style="width: 50%">
<col style="width: 16%">
<col style="width: 19%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">ref_group_ids</th>
<th style="text-align: left;">ref_path</th>
<th style="text-align: left;">Ref.primary.usage</th>
<th style="text-align: left;">Ref.contextual.Usage</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">abscessus</td>
<td style="text-align: left;">test/data/refs/mycobacterium_abscessus_reference1.fna</td>
<td style="text-align: left;">required</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">leprae</td>
<td style="text-align: left;">test/data/refs/mycobacterium_leprae_reference1.fna</td>
<td style="text-align: left;">required</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">leprae</td>
<td style="text-align: left;">test/data/refs/mycobacterium_leprae_reference2.fna</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">optional</td>
</tr>
<tr class="even">
<td style="text-align: left;">leprae</td>
<td style="text-align: left;">test/data/refs/mycobacterium_leprae_reference3.fna</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">optional</td>
</tr>
<tr class="odd">
<td style="text-align: left;">leprae</td>
<td style="text-align: left;">test/data/refs/mycobacterium_leprae_reference4.fna</td>
<td style="text-align: left;"></td>
<td style="text-align: left;">optional</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<p>Note that the “ref_group_ids” column in the sample input csv needs to match the sample csv:</p>
<div class="cell">
<div class="cell-output-display">
<div class="kable-table">
<table class="table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;">sample_id</th>
<th style="text-align: left;">ncbi.accession</th>
<th style="text-align: left;">ref_group_ids</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">mycobacterium_abscessus1</td>
<td style="text-align: left;">ERR7253671</td>
<td style="text-align: left;">abscessus</td>
</tr>
<tr class="even">
<td style="text-align: left;">mycobacterium_abscessus2</td>
<td style="text-align: left;">ERR7253669</td>
<td style="text-align: left;">abscessus</td>
</tr>
<tr class="odd">
<td style="text-align: left;">mycobacterium_abscessus3</td>
<td style="text-align: left;">ERR7253671</td>
<td style="text-align: left;">abscessus</td>
</tr>
<tr class="even">
<td style="text-align: left;">mycobacterium_leprae1</td>
<td style="text-align: left;">SRR6241707</td>
<td style="text-align: left;">leprae</td>
</tr>
<tr class="odd">
<td style="text-align: left;">mycobacterium_leprae2</td>
<td style="text-align: left;">SRR6241708</td>
<td style="text-align: left;">leprae</td>
</tr>
<tr class="even">
<td style="text-align: left;">mycobacterium_leprae3</td>
<td style="text-align: left;">SRR6241709</td>
<td style="text-align: left;">leprae</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<p>When enter the command to run the pipeline, the path this reference csv will need to be specified:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">nextflow</span> run nf-core/pathogensurveillance <span class="at">--sample_inut</span> mycobacterium_samples.csv <span class="at">--reference_input</span> mycobacterium_references.csv <span class="at">--output_dir</span> mycobacterium_test <span class="at">--download_bakta_db</span> true <span class="at">-profile</span> docker </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>